typescript-testing-examples/eval-in-python/README.md

```diff
@@ -0,0 +1,206 @@
+    )
+    all_feedback = batched_evaluator.batch(list(runs))
+    scores = [feedback.score for feedback in all_feedback if feedback.score is not None]
+    print(f"Average score: {sum(scores) / len(scores)}")
+
+
+if __name__ == "__main__":
+    dataset_name = "Structured Output Example"
+    project_name = os.environ.get("LANGCHAIN_PROJECT", "Unit Testing") + dataset_name
+    evaluate_existing_project(project_name=project_name)
+```
+
+The script does the following:
+1. Select the test project to evalute based on the same naming convention included in [index.ts](./src/index.ts)
+2. List the runs in the project
+3. Calls `client.evaluate_run` on each run, applying a custom evaluator (more info below) to generate the feedback.
+- (optional) We are using a RunnableLambda to help parallelize this call, but you could run it in a normal for loop.
+
+The custom evaluator is defined below:
+
+```python
+class JsonValueDistance(RunEvaluator):
+    def __init__(self) -> None:
+        super().__init__()
+        self.distance_evaluator = evaluation.load_evaluator("string_distance")
+        self.eval_name = "json_string_similarity"
+
+    def evaluate_run(
+        self, run: Run, example: Optional[Example] = None
+    ) -> EvaluationResult:
+        """Evaluate an example."""
+        if run.outputs is None:
+            raise ValueError("Run outputs cannot be None")
+        prediction = next(iter(run.outputs.values()))
+        reference = next(iter(example.outputs.values()))
+        if not reference:
+            return EvaluationResult(
+                key=self.eval_name, score=None, comment="No reference"
+            )
+        try:
+            obj = json.loads(prediction)
+        except Exception as e:
+            return EvaluationResult(key=self.eval_name, score=0.0, comment=str(e))
+        distances = []
+        for key in reference:
+            if key in obj:
+                norm_distance = self.distance_evaluator.evaluate_strings(
+                    prediction=str(obj[key]), reference=str(reference[key])
+                )["score"]
+            else:
+                norm_distance = 1.0
+            distances.append(norm_distance)
+        return EvaluationResult(
+            key=self.eval_name, score=1.0 - sum(distances) / len(distances)
+        )
+```
+
+This is meant to illustrate how you can apply custom logic to the traced run to generate a score. In this case, the evaluator checks for the average string distance for each value in a parsed json dictionary. A value of 1 means that all the reference keys have values that are approximately equal. A value of 0 means the json was invalid or the strings don't have overlap.
+
+Once the test script is executed, the evaluation results will be assigned to your JS test project! You can click on the "Test Runs" tab of the "Structured Output Example" dataset to view the results.
+
+
+(note: it may take up to a minute for the results to appear. Thank you for your patience!) 
```

Comment:
Will your concurrency fixes help with this @nfcampos?

Reply:
Not yet. We are looking to add eager project type assignment